arangod/Cluster/ClusterInfo.cpp

```diff
@@ -4232,6 +4288,83 @@ Result ClusterInfo::finishModifyingAnalyzerCoordinator(
   return Result(TRI_ERROR_NO_ERROR);
 }
 
+void ClusterInfo::initMetricsUpdate() {
+  VPackBuilder builderNewRebootId;
+  builderNewRebootId.add(VPackValue{kInvalidRebootId});
+  VPackBuilder builderNewServerId;
+  builderNewServerId.add(VPackValue{kInvalidServerId});
+
+  AgencyWriteTransaction const write{
+      {{kMetricsUpdateRebootId, AgencyValueOperationType::SET,
+        builderNewRebootId.slice()},
+       {kMetricsUpdateServerId, AgencyValueOperationType::SET,
+        builderNewServerId.slice()}},
+      {{kMetricsUpdateRebootId, AgencyPrecondition::Type::EMPTY, true},
+       {kMetricsUpdateServerId, AgencyPrecondition::Type::EMPTY, true}}};
+  AgencyComm ac{_server};
+  while (true) {
+    auto const r = ac.sendTransactionWithFailover(write);
+    if (r.successful() ||
+        r.httpCode() == rest::ResponseCode::PRECONDITION_FAILED) {
+      return;
+    }
+  }
+}
+
+bool ClusterInfo::lockMetricsUpdate() {
+  LOG_TOPIC("bfdb0", INFO, Logger::CLUSTER) << "Start lock metrics update";
+  auto ourRebootId = ServerState::instance()->getRebootId().value();
+  auto ourServerId = ServerState::instance()->getId();
+  TRI_ASSERT(ourRebootId != kInvalidRebootId &&
+             ourServerId != kInvalidServerId);
+  auto& cache = _server.getFeature<ClusterFeature>().agencyCache();
+  auto [lockRebootId, lockServerId] = readMetricsUpdate(cache);
+  if (lockRebootId != kInvalidRebootId || ourServerId != kInvalidServerId) {
+    _metricsUpdateGuard = _rebootTracker.callMeOnChange(
+        {lockServerId, RebootId{lockRebootId}},
+        [this, rebootId = lockRebootId, serverId = lockServerId] {
+          AgencyComm ac{_server};
+          changeMetricsUpdate(ac, rebootId, serverId, kInvalidRebootId,
+                              kInvalidServerId);
+        },
+        "Unlock MetricsUpdate if current owner die");
+    LOG_TOPIC("bfdb1", INFO, Logger::CLUSTER)
+        << "Cannot lock metrics update, it's already locked";
+    return false;
+  }
+  _metricsUpdateGuard.callAndClear();
+  AgencyComm ac{_server};
+  return changeMetricsUpdate(ac, kInvalidRebootId, kInvalidServerId,
+                             ourRebootId, ourServerId);
+}
+
+bool ClusterInfo::unlockMetricsUpdate() noexcept {
+  LOG_TOPIC("bfda0", INFO, Logger::CLUSTER) << "Start unlock metrics update";
+  AgencyComm ac{_server};
+  auto ourRebootId = ServerState::instance()->getRebootId().value();
+  auto ourServerId = ServerState::instance()->getId();
```

Comment:
this can't be `noexcept` as `getId()` returns `std::string` by value  ;-)

Reply:
I know and think std::terminate if bad alloc is ok